### PR TITLE
Framework: Enable Parallel Tests on Circle-CI

### PR DESCRIPTION
**Pull request by blowery**
Part of #3822.

This is an attempt to enable test and lint splitting on CircleCI, to take advantage of parallel test containers.

To really make this work, we need to drop the test configuration variable in the web settings, as that is running in addition to the split tests.

Early returns show that moving to 3 containers would drop overall test time to around 4m. Going to our full 8 containers should drop it to around 2m, even before all the test refactoring.

One problem with going full 8 is that the desktop app shares our container set and it seems to be constantly rebuilding itself, even with no changes to the code.

### CODE REVIEW

**Comment by blowery** [comment]
Might land #3794 first so we can see the actual benefit to parallel test runs

**Comment by blowery** [comment]
Nice, gotta love a `03:32` build. :)

File: circle.yml [hunk 1]
```diff
@@ -1,4 +1,17 @@
...
+test:
+  override:
+    - ./node_modules/.bin/eslint --quiet:
+        parallel: true
+        files:
+          - client/**/*.js
+          - client/**/*.jsx
+          - server/**/*.js
+          - server/**/*.jsx
+    - NODE_ENV=development make build && NODE_ENV=development ./bin/run-tests:
```
Comment (gziolo) [inline circle.yml]: In `run-tests` we overwrite `NODE_ENV`. It seems like the intention is to use development env.

File: client/my-sites/plugins/plugin-activate-toggle/test/index.jsx [hunk 2]
```diff
@@ -43,6 +43,7 @@ describe( 'PluginActivateToggle', function() {
 	} );
 
 	beforeEach( function() {
+		this.timeout( 10 * 1000 );
```
Comment (gziolo) [inline client/my-sites/plugins/plugin-activate-toggle/test/index.jsx]: 10 seconds seems to be a lot! It is 3 times longer than the new single test runner we work on atm :D

**Comment by gziolo** [comment]
`03:32` is great improvement comparing to `8:55` we have at the moment! :+1:

File: circle.yml (hunk 1 above)
Comment (blowery) [inline circle.yml]: This actually gets really confusing. Some Makefile ignore an already set NODE_ENV and reset it to TEST. And run-tests sets it, but doesn't export it or pass it along, so subprocs don't seem to pick it up. 

In the current system I think we're actually using the outer node_env just to generate the config file? Would be nice to not have the switcheroo and just use an env of test everywhere.

File: client/my-sites/plugins/plugin-activate-toggle/test/index.jsx (hunk 2 above)
Comment (blowery) [inline client/my-sites/plugins/plugin-activate-toggle/test/index.jsx]: Yeah 10 was a guess. I'd rather have this way too lenient since circle is unpredictably slow. Once we have one runner, we can bump up the global timeout only when running under circle and remove these timeout hacks.

Comment (gziolo) [inline client/my-sites/plugins/plugin-activate-toggle/test/index.jsx]: it sounds like a really good plan :+1:

File: circle.yml (hunk 1 above)
Comment (gziolo) [inline circle.yml]: That's get us to another unrelated change we discussed today with @nb. We should at some point take advantage of logic build by @mjangda around `config/_shared.json` file and extract config setting that are common for all envs.

File: circle.yml [hunk 3]
```diff
@@ -1,4 +1,17 @@
 machine:
   node:
     version: 4.3.0
-
+test:
+  override:
+    - ./node_modules/.bin/eslint --quiet:
```
Comment (aduth) [inline circle.yml]: Do we worry about the maintainability of splitting this off of the `lint` Makefile target?

(First time I've noticed `mixedindentlint` as a dependent target of `lint`)

Comment (blowery) [inline circle.yml]: There's no way to pass args to a makefile target (that i'm aware of?), so ... not sure how else to do it. maybe move to a `npm run lint` which runs eslint and move the makefile to use that?

File: circle.yml (hunk 1 above)
Comment (blowery) [inline circle.yml]: Yup, complete agreement. I was debating doing that here, but copying the dev config was simpler and less prone to introduce crazy things.

File: circle.yml (hunk 3 above)
Comment (aduth) [inline circle.yml]: >maybe move to a `npm run lint` which runs eslint and move the makefile to use that?

I'm a bit of the opinion this is the direction we should head anyways :smile: (Related: #3520)

Is `npm` any better about passing arguments to the scripted commands? In my own experiences, I seem to recall this being an issue.

Comment (blowery) [inline circle.yml]: No idea. Can find out.

Comment (blowery) [inline circle.yml]: http://stackoverflow.com/questions/11580961/sending-command-line-arguments-to-npm-script

Comment (blowery) [inline circle.yml]: In https://github.com/Automattic/wp-calypso/commit/ac95127d5bc62f06e9dde597eb5d7657b559badc

One odd bit: passing $(JS_FILES) through made the linter bomb. Unsure why.

**Comment by blowery** [comment]
OK, now `npm run lint` just works and is reasonably fast.

**Comment by blowery** [comment]
So, we good? I'd love to land this.

**Comment by aduth** [comment]
Confirmed `npm run lint`, `make lint`, `npm test`, and `make test` all exit with failing error code when errors present in linting or testing.

I'm personally still confused on a few bits, specifically around when `config/test.json` is used and why we're not passing `NODE_ENV=test` into `./bin/run-tests` script, but I think I'm failing to understand how the config is passed through those scripts.

I'll leave it to you to decide whether there's concerns there. Otherwise, this is looking good to go :+1:

**Comment by blowery** [comment]
> I'm personally still confused on a few bits, specifically around when config/test.json is used and why we're not passing NODE_ENV=test into ./bin/run-tests script, but I think I'm failing to understand how the config is passed through those scripts.

I think now that we have a test.json we can just use NODE_ENV=test and be fine. It didn't work in the past because there was no test config to pull from.

How it was working for the server tests is still a bit of a mystery to me.

**Comment by blowery** [comment]
OK, happy now. Landing.
